### DISCUSSION TITLE
In-memory proxy for caching data

### DISCUSSION THREAD

**Discussion by hamboomger** [simolus3/drift#1844] [Q&A] [answered]
Hello,
First and foremost - this library is insanely good, easily the best database client I've used in my 6 years of work as a web developer. Now for the actual topic of discussion - I also must preface that I am pretty new to the mobile development, so it might be stupid: _what are the good practices of using this library when the in-memory caching is involved as well?_

Let's say there is a screen with list of profiles, each time user goes to this screen, the app will show profiles from cache (via db call using drift), then it will actualise the data from server and update cache if anything changes. So far so good, but when the user goes to the aforementioned screen, there is a split second when the data from the db is still fetching and the page is empty. The gut reaction I have is to pre-load this data on the application start up, so that it will fetch the data from the memory instead of the database.

That is basically where my train of thought ends - I'm not sure what is the best way to approach this issue. My initial idea and the reason why I’m here is to make a proxy layer between the database and application, which will manage the in-memory caching logic. But it occurred to me that drift might already have some solution, or there is some similar discussion on this point, so why not ask the community/developers of drift :)

**Comment by simolus3** [comment, marked as answer]
> First and foremost - this library is insanely good, easily the best database client I've used in my 6 years of work as a web developer

Thanks! :heart: 

> The gut reaction I have is to pre-load this data on the application start up, so that it will fetch the data from the memory instead of the database.

This isn't what you asked for, but first a general note: For performance reasons alone, this shouldn't be necessary. sqlite3 will cache parts of your database in-memory as well. So for databases that aren't too large, you'll get most drift/sqlite3 queries answered from memory either way.

> So far so good, but when the user goes to the aforementioned screen, there is a split second when the data from the db is still fetching and the page is empty

Are you using a query stream for this? I assume that this is mostly due to the data being fetched asynchronously (e.g. if you use a `StreamBuilder`, it would first build an initial state and then very quickly update to the state with data). That's likely what's causing the flicker you're seeing with the initially-empty screen: It's just the fact that `Stream`/`StreamBuilder` are fundamentally async.

If you reach the screen with all the profiles through a tap, you could wait for the first snapshot before pushing the route. In your `StreamBuilder`, you could then use `initialData` to pass the initial snapshot and avoid an empty screen.

Drift has no builtin integration for caching. You could keep the stream for this query active from the start of your app and cache the results yourself. Bu be aware that having _too many_ query streams active when you don't need them may also slow your app down since drift will fetch results again for every potential update.

**Reply by hamboomger** [reply]
Oh wow, drift actually has internal in-memory cache as well, didn't know that! Thank you kind sir for a thorough response, I do actually use Streams and I think your solution should work well in my case. Tell me if there's something I can help with, otherwise I'll just stick around this project after I get a grasp of it and try to help any soul in dire need of help :)
